[LoopPeel] Peel last iteration to enable load widening

guy-david opened the pull request:
In loops that contain multiple consecutive small loads (e.g., 3 bytes loading i8's), peeling the last iteration makes it safe to read beyond the accessed region, enabling the use of a wider load (e.g., b32) for all other N-1 iterations.

Patterns such as:
```
  %a = load i8, ptr %p
  %b = load i8, ptr %p+1
  %c = load i8, ptr %p+2
  ...
  %p.next = getelementptr i8, ptr %p, 3
```

Can be transformed to:
```
  %wide = load b32, ptr %p     ; Read 4 bytes
  %frozen = freeze b32 %wide
  %value = bitcast b32 %frozen to i32
  %a = trunc
  %b = lshr + trunc
  %c = lshr + trunc
  %p.next = getelementptr i8, ptr %p, 3
  ...
```

The wide load is of a byte type rather than an integer to handle poison values correctly. Byte values carry poison per bit instead of per value, so the read-ahead bytes, which the peel makes dereferenceable, cannot taint the grouped bytes.

This acts as a fallback strategy when vectorization fails and has significant performance uplifts, most notably in image processing where processing an RGB buffer is common.

Assisted-by: Opus 4.8

guy-david committed:
[LoopPeel] Peel last iteration to enable load widening

In loops that contain multiple consecutive small loads (e.g., 3 bytes
loading i8's), peeling the last iteration makes it safe to read beyond
the accessed region, enabling the use of a wider load (e.g., i32) for
all other N-1 iterations.

Patterns such as:
```
  %a = load i8, ptr %p
  %b = load i8, ptr %p+1
  %c = load i8, ptr %p+2
  ...
  %p.next = getelementptr i8, ptr %p, 3
```

Can be transformed to:
```
  %wide = load i32, ptr %p  ; Read 4 bytes
  %a = trunc
  %b = lshr + trunc
  %c = lshr + trunc
  ...
```

This acts as a fallback strategy when vectorization fails and has
significant performance uplifts, most notably in image processing
where processing an RGB buffer is common.

nikic commented on llvm/test/Transforms/LoopUnroll/peel-last-iteration-load-widening.ll:
```diff
@@ -0,0 +1,616 @@
+  %p1 = getelementptr inbounds i8, ptr %p, i64 1
+  %p2 = getelementptr inbounds i8, ptr %p, i64 2
+
+  %a = load i8, ptr %p, align 4
```
How can the alignment here be 4 if this is a stride 3 pointer?

nikic reviewed (commented):
As a very high level question, why do we want to peel here instead of unroll?

For example, if we're accessing 3 elements, we could unroll twice and use a 4 + 2 byte load -- instead of two overlapping 4 byte loads.

Esp for the case of unrolling twice (where code size impact should be about the same as peeling), it seems like unrolling would be more beneficial, esp as it can also enable other optimizations?

Can you share any samples of code this is intended to optimize?

fhahn commented on llvm/lib/Transforms/Scalar/LoopUnrollPass.cpp:
```diff
@@ -1316,6 +1324,12 @@ tryToUnrollLoop(Loop *L, DominatorTree &DT, LoopInfo *LI, ScalarEvolution &SE,
+      // Widen consecutive loads after last-iteration peeling
+      if (PP.PeelLast) {
+        const DataLayout &DL = L->getHeader()->getDataLayout();
+        widenLoadsAfterPeel(*L, SE, DL, TTI, DT);
```
no need to pass DL, can get in the caller as well directly from L

fhahn commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -746,13 +751,148 @@ static bool violatesLegacyMultiExitLoopCheck(Loop *L) {
+  Value *BasePtr;
+  // First load instruction in the program order.
+  LoadInst *FirstLoad;
+  // Pairs of (load instruction, offset from base).
```
```suggestion
  // Pairs of (load instruction, offset from base) sorted by offset.
```

fhahn commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -746,13 +751,148 @@ static bool violatesLegacyMultiExitLoopCheck(Loop *L) {
+    return 0;
+  if (!L.isInnermost())
+    return 0;
+  if (!canPeelLastIteration(L, SE))
```
can we combine this with the other peeling-from-last-iteration logic to avoid repated calls ?

fhahn commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -746,13 +751,148 @@ static bool violatesLegacyMultiExitLoopCheck(Loop *L) {
+
+// Returns 1 if peeling the last iteration would enable widening load groups to
+// natural sizes. Returns 0 otherwise.
+static unsigned peelLastForLoadWidening(Loop &L, ScalarEvolution &SE,
```
clearer to return true?

guy-david committed:
Address comments 1

guy-david commented:
ping :)

fhahn commented on llvm/lib/Transforms/Scalar/LoopUnrollPass.cpp:
```diff
@@ -1025,17 +1025,15 @@ shouldPartialUnroll(const unsigned LoopSize, const unsigned TripCount,
+    const unsigned MaxTripCount, const bool MaxOrZero,
+    const unsigned TripMultiple, const UnrollCostEstimator &UCE,
+    TargetTransformInfo::UnrollingPreferences &UP,
+    TargetTransformInfo::PeelingPreferences &PP, bool AllowLoadWideningPeel) {
```
I think AllowLoadWideningPeel would fit better & cleaner in PeelingPreferneces.

fhahn commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -749,13 +753,194 @@ static bool violatesLegacyMultiExitLoopCheck(Loop *L) {
+      Instruction::Load, WideType, Align(1), AddrSpace, CostKind);
+  for (const auto &[Load, Offset] : Loads) {
+    unsigned LoadBits = Load->getType()->getScalarSizeInBits();
+    uint64_t BitPosition = (Offset.getSExtValue() - FirstOffset) * 8;
```
I think we need to either avoid getSExtValue or guard it fitting into i64, otherwise we crash with something like

```
target datalayout = "e-p:128:128:128-i64:64-i128:128"

define void @test_huge_offset(ptr %src, ptr %dst, i32 %n) {
entry:
  br label %loop

loop:
  %i = phi i32 [ 0, %entry ], [ %i.next, %loop ]
  %p = phi ptr [ %src, %entry ], [ %p.next, %loop ]

  ; The offset here (1e20) needs more than 64 bits to represent.
  %pbig = getelementptr inbounds i8, ptr %p, i128 100000000000000000000

  %a = load i8, ptr %p, align 1
  %b = load i8, ptr %pbig, align 1

  %sum = add i8 %a, %b
  %dst.i = getelementptr inbounds i8, ptr %dst, i32 %i
  store i8 %sum, ptr %dst.i, align 1

  %p.next = getelementptr inbounds i8, ptr %p, i128 3
  %i.next = add i32 %i, 1
  %cond = icmp ne i32 %i.next, %n
  br i1 %cond, label %loop, label %exit

exit:
  ret void
}
```

fhahn commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -749,13 +753,194 @@ static bool violatesLegacyMultiExitLoopCheck(Loop *L) {
+                                        Load->getAlign(), AddrSpace, CostKind);
+  }
+
+  // Cost of wide load + extraction operations (shift + trunc for each load).
```
We have something very similar in VectorCombine (`scalarizeExtExtract`) I think. may be worth sharing some of the logic if possible, although not sure where the best place would be

guy-david replied:
I thought about adding a helper in DataLayout but it couldn't find a clean refactor, mainly because `scalarizeExtExtract` uses a vector type while here there's none.

fhahn commented on llvm/test/Transforms/LoopUnroll/AArch64/peel-last-iteration-load-widening.ll:
```diff
@@ -0,0 +1,1694 @@
+; CHECK:       [[LOOP]]:
+; CHECK-NEXT:    [[I:%.*]] = phi i32 [ 0, %[[ENTRY_SPLIT]] ], [ [[I_NEXT:%.*]], %[[LOOP]] ]
+; CHECK-NEXT:    [[P:%.*]] = phi ptr [ [[SRC]], %[[ENTRY_SPLIT]] ], [ [[P_NEXT:%.*]], %[[LOOP]] ]
+; CHECK-NEXT:    [[TMP2:%.*]] = load i32, ptr [[P]], align 4
```
I think we need to be careful regarding poison here. If any of the 3 loads is poison, all 3 extracted values will be poison I think (https://alive2.llvm.org/ce/z/xhMUTS). For this test, it is fine, but it should be possible to add a test where this would be incorrect (e.g. passing the loaded values to separate function calls)


Would be good to check the logic  in `scalarizeExtExtract`

I think we also need to be careful about the `align`. I don't think the new load is guaranteed to be aligned by 4? You probably need to set the correct alignment at load construction

guy-david replied:
I have come to realize that the issue with poison is even greater with regards to the additional, now dereferenceable, bits being loaded. For the already dereferenceable bits we could add a guard, but it's much harder to reason noundef / poison for the additional bits because subsequent store operations in the loop may write to them.
Therefore, I'm looking into combining several loads into a single scalar load that has the same amount of total bits (no rounding to the next power of 2) to avoid this issue and add an `assume` for dereferenceability for a later patch (in codegen) to pick up.

fhahn replied:
https://github.com/llvm/llvm-project/pull/179642 should probably help here

artagnon commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -895,6 +1098,70 @@ void llvm::computePeelCount(Loop *L, unsigned LoopSize,
+          Extracted = Builder.CreateLShr(
+              Extracted, ConstantInt::get(WideLoad->getType(), BitOffset));
+        if (LoadBits < SizeInBits)
+          Extracted = Builder.CreateTrunc(Extracted, Load->getType());
```
Hm, I wonder if it's possible to use llvm.pext here?

guy-david replied:
Because the bits are contiguous, InstCombine canonicalizes it into a shift anyway.

artagnon commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -895,6 +1098,70 @@ void llvm::computePeelCount(Loop *L, unsigned LoopSize,
+      LoadInst *InsertPoint = Group.FirstLoad;
+      IRBuilder<> Builder(InsertPoint);
+      Value *BasePtr = Group.BasePtr;
+      int64_t FirstOffset = Group.Loads[0].second.getSExtValue();
```
Hm, I wonder if it would make sense to simply store an int64_t to begin with?

artagnon commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -749,13 +753,194 @@ static bool violatesLegacyMultiExitLoopCheck(Loop *L) {
+  for (const auto &[Load, Offset] : Loads) {
+    if (Load->getPointerAddressSpace() != AddrSpace)
+      return std::nullopt;
+  }
```
```suggestion
  for (any_of(make_first_range(Loads), [AddrSpace] (LoadInst *Load) { return Load->getPointerAddressSpace() != AddrSpace; }))
      return std::nullopt;
  }
```

Make me wonder why we even insert loads in different address space into the map in the first place?

artagnon commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -749,13 +753,194 @@ static bool violatesLegacyMultiExitLoopCheck(Loop *L) {
+    return std::nullopt;
+  int64_t StepVal = ConstStep->getValue()->getSExtValue();
+  if (StepVal != static_cast<int64_t>(TotalBytes))
+    return std::nullopt;
```
```suggestion
  if (!match(SE.getSCEV(Base),
             m_scev_AffineAddRec(m_SCEV(), m_scev_SpecificSInt(TotalBytes),
                                 m_SpecificLoop(&L))))
    return std::nullopt;
```

artagnon commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -749,13 +753,194 @@ static bool violatesLegacyMultiExitLoopCheck(Loop *L) {
+  // Sort loads by offset first since we need this for cost calculation.
+  llvm::sort(Loads, [](const auto &A, const auto &B) {
+    return A.second.slt(B.second);
+  });
```
```suggestion
  llvm::sort(Loads);
```

Redundant if we store int64_t in the first place?

artagnon commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -895,6 +1098,70 @@ void llvm::computePeelCount(Loop *L, unsigned LoopSize,
+        findLoadGroupsForWidening(BB, L, SE, DL, TTI);
+    for (const LoadGroup &Group : Groups) {
+      LoadInst *InsertPoint = Group.FirstLoad;
+      IRBuilder<> Builder(InsertPoint);
```
```suggestion
      IRBuilder<> Builder(Group.FirstLoad);
```

artagnon commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -895,6 +1098,70 @@ void llvm::computePeelCount(Loop *L, unsigned LoopSize,
+      }
+      // Delete the original loads.
+      for (auto &[Load, Offset] : Group.Loads)
+        Load->eraseFromParent();
```
```suggestion
      for_each(make_first_range(Group.Loads), [](LoadInst *Load) { Load->eraseFromParent; });
```

guy-david replied:
Formatter actually changes this back into a regular for loop.

artagnon commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -749,13 +753,194 @@ static bool violatesLegacyMultiExitLoopCheck(Loop *L) {
+    return false;
+  BasicBlock *Latch = L.getLoopLatch();
+  if (!Latch)
+    return false;
```
```suggestion
  BasicBlock *Latch = L.getLoopLatch();
  if (!EnablePeelForLoadWidening || !L.isInnermost() || !Latch)
    return false;
```

artagnon commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -749,13 +753,194 @@ static bool violatesLegacyMultiExitLoopCheck(Loop *L) {
+      // Skip volatile/atomic loads. Also skip non-integer types since the
+      // widening uses integer shift/truncate operations to extract bytes.
+      // Supporting floats would require additional bitcasts and consecutive
+      // float loads are less probable to create odd amounts of bytes.
```
I think using llvm.pext would justify this?

artagnon commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -749,13 +753,194 @@ static bool violatesLegacyMultiExitLoopCheck(Loop *L) {
+          !Load->getType()->isIntegerTy())
+        continue;
+      Value *Ptr = Load->getPointerOperand();
+      APInt Offset(DL.getIndexTypeSizeInBits(Ptr->getType()), 0);
```
```suggestion
      APInt Offset(DL.getIndexTypeSizeInBits(Ptr->getType()), AddressSpace);
```

I think we have the AddressSpace somewhere? Perhaps worth storing?

guy-david committed:
[NFC][LoopPeel] Add tests for load widening

guy-david committed:
[LoopPeel] Peel last iteration to enable load widening

In loops that contain multiple consecutive small loads (e.g., 3 bytes
loading i8's), peeling the last iteration makes it safe to read beyond
the accessed region, enabling the use of a wider load (e.g., b32) for
all other N-1 iterations.

Patterns such as:
```
  %a = load i8, ptr %p
  %b = load i8, ptr %p+1
  %c = load i8, ptr %p+2
  ...
  %p.next = getelementptr i8, ptr %p, 3
```

Can be transformed to:
```
  %wide = load b32, ptr %p     ; Read 4 bytes
  %frozen = freeze b32 %wide
  %value = bitcast b32 %frozen to i32
  %a = trunc
  %b = lshr + trunc
  %c = lshr + trunc
  %p.next = getelementptr i8, ptr %p, 3
  ...
```

The wide load is of a byte type rather than an integer to handle poison
values correctly. Byte values carry poison per bit instead of per value,
so the read-ahead bytes, which the peel makes dereferenceable, cannot
taint the grouped bytes.

This acts as a fallback strategy when vectorization fails and has
significant performance uplifts, most notably in image processing
where processing an RGB buffer is common.

Assisted-by: Opus 4.8
